engine/table/src/main/java/io/deephaven/engine/table/impl/by/GroupByChunkedOperator.java

```diff
@@ -331,15 +439,123 @@ public ModifiedColumnSet apply(@NotNull final ModifiedColumnSet upstreamModified
+                 final ResettableWritableObjectChunk<RowSetBuilderRandom, Values> removedBuildersResettableChunk =
+                         ResettableWritableObjectChunk.makeResettableChunk();
+                 final RowSequence.Iterator destinationsIterator =
+                         stepDestinations.getRowSequenceIterator()) {
+
+                // noinspection unchecked
+                final WritableObjectChunk<WritableRowSet, Values> rowSetBackingChunk =
+                        rowSetResettableChunk.asWritableObjectChunk();
+                // noinspection unchecked
+                final WritableObjectChunk<RowSetBuilderRandom, Values> addedBuildersBackingChunk =
+                        addedBuildersResettableChunk.asWritableObjectChunk();
+                // noinspection unchecked
+                final WritableObjectChunk<RowSetBuilderRandom, Values> removedBuildersBackingChunk =
+                        removedBuildersResettableChunk.asWritableObjectChunk();
+
+                while (destinationsIterator.hasMore()) {
+                    final long firstSliceDestination = destinationsIterator.peekNextKey();
+                    final long firstBackingChunkDestination =
+                            rowSets.resetWritableChunkToBackingStore(rowSetResettableChunk, firstSliceDestination);
+                    addedBuilders.resetWritableChunkToBackingStore(addedBuildersResettableChunk, firstSliceDestination);
+                    removedBuilders.resetWritableChunkToBackingStore(removedBuildersResettableChunk, firstSliceDestination);
+
+                    final long lastBackingChunkDestination =
+                            firstBackingChunkDestination + rowSetBackingChunk.size() - 1;
+                    final RowSequence initialDestinationsSlice =
+                            destinationsIterator.getNextRowSequenceThrough(lastBackingChunkDestination);
+
+                    initialDestinationsSlice.forAllRowKeys((final long destination) -> {
+                        final int backingChunkOffset =
+                                Math.toIntExact(destination - firstBackingChunkDestination);
+                        try (final WritableRowSet addRowSet =
+                                nullToEmpty(extractAndClearBuilderRandom(addedBuildersBackingChunk, backingChunkOffset));
+                             final WritableRowSet removeRowSet =
+                                nullToEmpty(extractAndClearBuilderRandom(removedBuildersBackingChunk, backingChunkOffset))) {
+                            final WritableRowSet workingRowSet = rowSetForSlot(backingChunkOffset);
+
+                            // process removes
+                            final boolean wasNonEmpty = removeRowSet.isNonempty();
+                            workingRowSet.remove(removeRowSet);
+                            if (wasNonEmpty && workingRowSet.isEmpty()) {
+                                onEmptied(destination);
```

Comment:
This is too late in update processing to meaningfully report emptied to reincarnated. We use that to determine `downstream` before this call is made. If we can't determine emptied or reincarnated correctly in the `addChunk` and `removeChunk` methods (and I don't think we can), we should just remove `extends BasicStateChangeRecorder` from the class signature.